controllers/chat/getSignedChannelList.js

```diff
@@ -0,0 +1,78 @@
+const {responseSuccess} = require('../../utils/Responses');
+const GetstreamSingleton = require('../../vendor/getstream/singleton');
+
+const MAX_ITERATIONS = 10;
+
+const __queryChannelBuilder = async (userId, limit, offset) => {
+  const client = GetstreamSingleton.getChatInstance();
+  const channels = await client.queryChannels({members: {$in: [userId]}}, [{last_message_at: -1}], {
+    limit: limit,
+    offset: offset,
+    state: true
+  });
+
+  return channels;
+};
+
+const __transformChannelData = (channel) => {
+  const newChannel = {...channel.data};
+  delete newChannel.config;
+  delete newChannel.own_capabilities;
+
+  const members = [];
+  Object.keys(channel.state.members).forEach((member) => {
+    members.push(channel?.state?.members[member]);
+  });
+
+  return {
+    ...newChannel,
+    members,
+    unreadCount: channel.state.unreadCount,
+    messages: channel.state.messages
+  };
+};
+
+const getSignedChannelList = async (req, res) => {
+  const {userId} = req;
+  const {limit = 100, offset = 0} = req.query;
+
+  let totalFetched = 0;
+  const queriedChannels = [];
+  const promisedChannels = [];
+
+  const client = GetstreamSingleton.getChatInstance();
+  try {
+    await client.connectUser({id: req.userId}, req.token);
+
+    while (totalFetched < limit) {
+      if (totalFetched > MAX_ITERATIONS * 30) throw new Error('Too many iterations');
+
+      const batchSize = Math.min(30, limit - totalFetched);
+      promisedChannels.push(
+        __queryChannelBuilder(userId, batchSize, Number(totalFetched) + Number(offset))
+      );
+      totalFetched += batchSize;
+      if (totalFetched >= limit) {
+        break;
+      }
+    }
+
+    const responses = await Promise.all(promisedChannels);
+    responses.forEach((response) => {
+      queriedChannels.push(...response);
+    });
+
+    const channels = queriedChannels.map(__transformChannelData);
+    return res.status(200).json(responseSuccess('Success retrieve channels', channels));
+  } catch (error) {
+    return res.status(error.statusCode ?? error.status ?? 400).json({
+      status: 'error',
+      code: error.statusCode ?? error.status ?? 400,
+      message: error.message
+    });
+  } finally {
+    await client.disconnectUser();
+  }
+};
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The `getSignedChannelList` function is not validating the input parameters `limit` and `offset`. It's important to validate these inputs to prevent potential issues such as SQL injection or fetching more data than the system can handle. Consider adding a middleware for input validation before this function is called.

```diff
const getSignedChannelList = async (req, res) => {
+ if (!Number.isInteger(limit) || limit < 1 || limit > MAX_LIMIT) {
+   return res.status(400).json({message: 'Invalid limit'});
+ }
+ if (!Number.isInteger(offset) || offset < 0) {
+   return res.status(400).json({message: 'Invalid offset'});
+ }

  const {userId} = req;
  const {limit = 100, offset = 0} = req.query;

  let totalFetched = 0;
  const queriedChannels = [];
  const promisedChannels = [];

  const client = GetstreamSingleton.getChatInstance();
  try {
    await client.connectUser({id: req.userId}, req.token);

    while (totalFetched < limit) {
      if (totalFetched > MAX_ITERATIONS * 30) throw new Error('Too many iterations');

      const batchSize = Math.min(30, limit - totalFetched);
      promisedChannels.push(
        __queryChannelBuilder(userId, batchSize, Number(totalFetched) + Number(offset))
      );
      totalFetched += batchSize;
      if (totalFetched >= limit) {
        break;
      }
    }

    const responses = await Promise.all(promisedChannels);
    responses.forEach((response) => {
      queriedChannels.push(...response);
    });

    const channels = queriedChannels.map(__transformChannelData);
    return res.status(200).json(responseSuccess('Success retrieve channels', channels));
  } catch (error) {
    return res.status(error.statusCode ?? error.status ?? 400).json({
      status: 'error',
      code: error.statusCode ?? error.status ?? 400,
      message: error.message
    });
  } finally {
    await client.disconnectUser();
  }
};

module.exports = getSignedChannelList;
```


<!-- This is an auto-generated comment by OSS CodeRabbit -->

Reply:
limit and offset type are already defined on the body validation middleware.
max fetching number has been defined on max iteration variable